### PR TITLE
Event Callback Decorator Implementation

### PR DESCRIPTION
**Pull request by burzum**
This PR will allow you to implement event callback decorators to give you a more fine grained control to filter them conditionally using the decorators.

This should resolve:
- #8779 RFC: Event filters
- #8657 Better before filters and conditional event listeners

### CODE REVIEW

File: tests/TestCase/Event/Decorator/ConditionDecoratorTest.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    public function testCanTriggerIf()
+    {
+        $callable = function(Event $event) {
+            return 'success';
+        };
+
+        $decorator = new ConditionDecorator($callable, [
+            'if' => function(Event $event) {
+                if (isset($event->data['canTrigger'])) {
```
Comment (markstory) [inline tests/TestCase/Event/Decorator/ConditionDecoratorTest.php]:
```php
return isset($event->data['canTrigger']);
```

File: tests/TestCase/Event/Decorator/ConditionDecoratorTest.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        $this->assertTrue($decorator->canTrigger($event));
+
+        $result = $decorator($event);
+        $this->assertEquals('success', $result);
+    }
+
+    /**
+     * testCallableRuntimeException
+     *
+     * @expectedException \RuntimeException
```
Comment (markstory) [inline tests/TestCase/Event/Decorator/ConditionDecoratorTest.php]: Since there are multiple exceptions, asserting the expected message can help ensure you're reaching the code you want.

File: src/Event/EventManager.php [hunk 3]
```diff
@@ -157,14 +158,14 @@ public function on($eventKey = null, $options = [], $callable = null)
...
-                'callable' => $options
+                'callable' => new FilterDecorator($options),
             ];
             return;
         }
         if ($argCount === 3) {
             $priority = isset($options['priority']) ? $options['priority'] : static::$defaultPriority;
             $this->_listeners[$eventKey][$priority][] = [
-                'callable' => $callable
+                'callable' => new FilterDecorator($callable, $options)
```
Comment (markstory) [inline src/Event/EventManager.php]: Could we get integration tests at the EventManager level?

File: src/Event/Decorator/ConditionDecorator.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    protected function _evaluateCondition($condition, Event $event)
+    {
+        if (!isset($this->_options[$condition])) {
+            if ($condition === 'unless') {
+                return false;
+            }
+            return true;
+        }
+        if (!is_callable($this->_options[$condition])) {
```
Comment (sukihub) [inline src/Event/Decorator/ConditionDecorator.php]: Would it be possible to have an array of conditions?

File: src/Event/Decorator/SubjectFilterDecorator.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ */
+namespace Cake\Event\Decorator;
+
+use Cake\Event\Event;
+use RuntimeException;
+
+/**
+ * Event Subject Filter Decorator
+ */
+class SubjectFilterDecorator extends BaseDecorator
```
Comment (sukihub) [inline src/Event/Decorator/SubjectFilterDecorator.php]: Can you maybe give me an example when to use this? I don't really see why this:
```php
$this->eventManager()->on('event', new SubjectFilterDecorator($callable, ['allowedSubject' => '\Class\Name']));
```
is better than:
```php
$this->eventManager()->on('event', ['if' => function ($event) { return $event->subject instanceof \Class\Name; }], $callable);
```

File: src/Event/EventManager.php (hunk 3 above)
Comment (sukihub) [inline src/Event/EventManager.php]: And do we want to use FilterDecorator for all the events, or only in case when `if` or `unless` options are present?

File: src/Event/Decorator/ConditionDecorator.php (hunk 4 above)
Comment (burzum) [inline src/Event/Decorator/ConditionDecorator.php]: Why would this be useful? Use cases?

File: src/Event/Decorator/SubjectFilterDecorator.php (hunk 5 above)
Comment (burzum) [inline src/Event/Decorator/SubjectFilterDecorator.php]: Convenience.

File: tests/TestCase/Event/Decorator/ConditionDecoratorTest.php (hunk 2 above)
Comment (burzum) [inline tests/TestCase/Event/Decorator/ConditionDecoratorTest.php]: Done.

File: tests/TestCase/Event/Decorator/ConditionDecoratorTest.php (hunk 1 above)
Comment (burzum) [inline tests/TestCase/Event/Decorator/ConditionDecoratorTest.php]: Done.

File: src/Event/Decorator/ConditionDecorator.php (hunk 4 above)
Comment (sukihub) [inline src/Event/Decorator/ConditionDecorator.php]: You can wrap them in a closure, but something like this:
```php
->on('Controller.startup', ['if' => [$isEditAction, $isNotOwner]], $denyAccess)
```
gives you simpler (resuable?) closures, communicates the intent better, is easy to write, etc.

Comment (burzum) [inline src/Event/Decorator/ConditionDecorator.php]: Might be true but I would like to keep the implementation as simple as possible for now, that's also a reason I've added only if and unless, @markstory told me to do so. This can be extended in another PR when this one is done.

File: src/Event/Decorator/SubjectFilterDecorator.php (hunk 5 above)
Comment (sukihub) [inline src/Event/Decorator/SubjectFilterDecorator.php]: I see how extending decorators/callables could be useful, but in this particular case, not really easier to read or write. And I've never needed to check event's subject so it's hard for me to imagine a use case.

Comment (burzum) [inline src/Event/Decorator/SubjectFilterDecorator.php]: Especially in pure listeners that are not behaviours, that deal with table objects, knowing the subject is pretty useful.

File: src/Event/Decorator/BaseDecorator.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * Calls a callable with the passed arguments.
+     *
+     * @param callable $callable The callable.
+     * @param array $args Arguments for the callable.
+     * @return mixed
+     */
+    public function call(callable $callable, $args)
```
Comment (jrbasso) [inline src/Event/Decorator/BaseDecorator.php]: @burzum Seems this method is not used anywhere else. Why it needs to be public if usually the trigger is done via the invoke? It almost seems like an static method that doesn't depend from the rest of the class.

I would suggest you to make this method protected and remove the callable parameter and using the `$this->_callable` directly. On classes that extend this base class it can be called directly without calling the `__invoke`.

File: src/Event/Decorator/BaseDecorator.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP Project
+ * @since         3.3.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Event\Decorator;
+
+class BaseDecorator
```
Comment (jrbasso) [inline src/Event/Decorator/BaseDecorator.php]: @burzum Usually base classes are abstract since they are don't have much value by itself. Seems your "child" classes always need to implement the `canTrigger` method. You can also make this method abstract here and move the logic inside `__invoke` from the other classes to here.

Makes sense or do you see other decorators not using this method?

File: src/Event/Decorator/ConditionDecorator.php (hunk 4 above)
Comment (markstory) [inline src/Event/Decorator/ConditionDecorator.php]: Without a list of callables, one can still use functional composition to 'join' multiple conditions together.

```php
->on('Controller.startup', [
  'if' => function ($event) use ($isEditAction, $isNotOwner) {
  return $isEditAction($event) && $isNotOwner($event);
  }
], $listener);
```

If we find this happening often, I bet we can add helpers that let you do things like

```php
// Must pass all callbacks
$condition = FunctionBuilder::all($isEditAction, $isNotOwner);

// Must pass at least one callback
$condition = FunctionBuilder::any(isEditAction, $isNotOwner);
```

File: src/Event/Decorator/BaseDecorator.php (hunk 6 above)
Comment (markstory) [inline src/Event/Decorator/BaseDecorator.php]: I agree with @jrbasso, this method doesn't need to be public.

It might be worth doing a quick profiling check to make sure this 'optimization' is still relevant.

File: src/Event/Decorator/ConditionDecorator.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            throw new RuntimeException(self::class . ' the `' . $condition . '` condition is not a callable!');
```
Comment (markstory) [inline src/Event/Decorator/ConditionDecorator.php]: Be careful here, you could get hit with an array to string conversion error. It would be safer to use `var_export()`

File: src/Event/Decorator/BaseDecorator.php (hunk 7 above)
Comment (burzum) [inline src/Event/Decorator/BaseDecorator.php]: > Usually base classes are abstract since they are don't have much value by itself. 

Agreed and already done.

> Makes sense or do you see other decorators not using this method?

Honestly, I have no idea. :) This is something time will show.

File: src/Event/Decorator/ConditionDecorator.php (hunk 8 above)
Comment (burzum) [inline src/Event/Decorator/ConditionDecorator.php]: Sorry, I can't follow you here? Why can I get an array here and how would var_export show this in a nice way (one per line)?

File: src/Event/Decorator/AbstractDecorator.php [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+     */
+    public function __invoke()
+    {
+        return $this->_call($this->_callable, func_get_args());
+    }
+
+    /**
+     * Calls the decorated callable with the passed arguments.
+     *
+     * @param callable $callable The callable.
```
Comment (jrbasso) [inline src/Event/Decorator/AbstractDecorator.php]: @burzum You don't have this param anymore.

File: src/Event/Decorator/AbstractDecorator.php [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+    /**
+     * Invoke
+     *
+     * @link http://php.net/manual/en/language.oop5.magic.php#object.invoke
+     * @return mixed
+     */
+    public function __invoke()
+    {
+        return $this->_call($this->_callable, func_get_args());
```
Comment (jrbasso) [inline src/Event/Decorator/AbstractDecorator.php]: @burzum You shouldn't pass the first param since the `_call` just expect the args now.

File: src/Event/Decorator/BaseDecorator.php (hunk 7 above)
Comment (jrbasso) [inline src/Event/Decorator/BaseDecorator.php]: @burzum What do you think about the part about the `canTrigger`? It would remove the `__invoke` on the children classes.

File: src/Event/Decorator/AbstractDecorator.php (hunk 10 above)
Comment (ADmad) [inline src/Event/Decorator/AbstractDecorator.php]: The `__invoke()` method isn't covered by a test which is why this mistake wasn't caught.

File: src/Event/Decorator/ConditionDecorator.php (hunk 8 above)
Comment (markstory) [inline src/Event/Decorator/ConditionDecorator.php]: Nevermind. I'm dumb. I misread and thought that `$this->_options[$condition]` was going into the error message.

File: src/Event/Decorator/BaseDecorator.php (hunk 7 above)
Comment (burzum) [inline src/Event/Decorator/BaseDecorator.php]: @jrbasso I'm not sure. Would be a solution but would imply as well that each decorator is implementing some kind of check. @markstory what are your thoughts on this topic?

Comment (markstory) [inline src/Event/Decorator/BaseDecorator.php]: Personally I only add base classes once I have 2 or 3 implementations, so I can be sure I have the generalizations right.

**Comment by burzum** [comment]
@markstory friendly ping. :smile:

**Comment by burzum** [comment]
@markstory anything else left on this I need to get done?

**Comment by markstory** [comment]
I'll take a look this week. I've been a bit underwater lately.

File: src/Event/Decorator/ConditionDecorator.php [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * @inheritdoc
+     * @return mixed
+     */
+    public function __invoke()
+    {
+        $args = func_get_args();
+        if (!$this->canTrigger($args[0])) {
+            return false;
```
Comment (markstory) [inline src/Event/Decorator/ConditionDecorator.php]: Won't this stop the event and prevent other listeners from being triggered? I would expect the condition/subject check to just make the decorated callable into a no-op not stop the event.

Comment (burzum) [inline src/Event/Decorator/ConditionDecorator.php]: Fixed and tested.

**Comment by burzum** [comment]
@markstory everything OK?
